tests/phpunit/tests/media.php

```diff
@@ -5173,64 +5405,67 @@ public function test_wp_get_loading_optimization_attributes_low_res_image() {
+		// Shortcodes processed outside of content blobs like 'the_content' always get `loading="lazy"`.
 		$this->assertSame(
-			$expected,
+			array(
+				'loading' => 'lazy',
+			),
 			wp_get_loading_optimization_attributes( 'img', $attr, 'do_shortcode' ),
-			$message
+			'Lazy-loading not applied to shortcodes outside the loop.'
 		);
 	}
 
-	public function data_wp_get_loading_optimization_attributes_in_shortcodes() {
-		return array(
-			'main_shortcode_image_should_have_fetchpriority_high'  => array(
-				'setup'    => function () {
-					global $wp_query;
+	/**
+	 * Tests that the `do_shortcode` context does not result in loading optimization changes when used within a content
+	 * blob.
+	 *
+	 * @ticket 58853
+	 * @dataProvider data_get_filters_with_do_shortcode_callback
+	 *
+	 * @covers ::wp_get_loading_optimization_attributes
+	 */
+	public function test_wp_get_loading_optimization_attributes_in_shortcodes_within_content_blob( $filter_name ) {
+		$result = null;
 
-					// Set WP_Query to be in the loop and the main query.
-					$wp_query->in_the_loop = true;
-					$this->set_main_query( $wp_query );
-				},
-				'expected' => array(
-					'fetchpriority' => 'high',
-				),
-				'message'  => 'Fetch priority not applied to during shortcode rendering.',
-			),
-			'main_shortcode_image_after_threshold_is_loading_lazy' => array(
-				'setup'    => function () {
-					global $wp_query;
+		remove_all_filters( $filter_name );
+		add_filter(
+			$filter_name,
+			function( $content ) use ( &$result ) {
+				$attr   = $this->get_width_height_for_high_priority();
+				$result = wp_get_loading_optimization_attributes( 'img', $attr, 'do_shortcode' );
+				return $content;
+			}
+		);
+		apply_filters( $filter_name, '' );
 
-					// Set WP_Query to be in the loop and the main query.
-					$wp_query->in_the_loop = true;
-					$this->set_main_query( $wp_query );
+		// Shortcodes processed within content blobs like 'the_content' should never get any loading optimization attributes.
+		$this->assertSame(
+			array(),
+			$result,
+			'Loading optimization unexpectedly applied to shortcodes within content blob.'
+		);
+	}
 
-					// Set internal flags so lazy should be applied.
-					wp_high_priority_element_flag( false );
-					wp_increase_content_media_count( 3 );
-				},
-				'expected' => array(
-					'loading' => 'lazy',
-				),
-				'message'  => 'Lazy-loading not applied to during shortcode rendering.',
-			),
-			'shortcode_image_outside_of_the_loop_are_loaded_lazy'  => array(
-				'setup'    => function () {
-					// Avoid setting up the WP_Query object for the loop.
-					return;
-				},
-				'expected' => array(
-					'loading' => 'lazy',
-				),
-				'message'  => 'Lazy-loading not applied to shortcodes outside the loop.',
-			),
+	/**
+	 * Gets filters for content blobs that by default have a `do_shortcode()` callback.
+	 *
+	 * @return array[]
+	 */
+	public function data_get_filters_with_do_shortcode_callback() {
+		return array(
```

Comment:
We can use this helper method to make this use named datasets while also making the datasets easier to read/add/modify.

```php
return self::text_array_to_dataprovider(
	array(
		'the_content',
		'widget_text_content',
		'widget_block_content',
	)
);
```

Reply:
Updated in https://github.com/WordPress/wordpress-develop/pull/4973/commits/79ebdddfceec96dec01e20c0bcad8be8fb653b82